### PR TITLE
[extractor/cspan] Support of C-Span congress videos

### PR DESCRIPTION
**Pull request by Grabien**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

A new extractor for C-Span congress videos:
https://www.c-span.org/congress/?chamber=house&date=2017-12-13&t=1513208380

### CODE REVIEW

File: yt_dlp/extractor/cspan.py [hunk 1]
```diff
@@ -242,3 +243,49 @@ def get_text_attr(d, attr):
...
+            'ext': 'mp4'
+        }
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._generic_id(url)
+        webpage = self._download_webpage(url, display_id)
+        file = open("sample.txt", "w")
+        file.write(webpage)
+        file.close()
```
Comment (MinePlayersPE) [inline yt_dlp/extractor/cspan.py:266]:
```suggestion
```

File: yt_dlp/extractor/cspan.py [hunk 2]
```diff
@@ -242,3 +243,49 @@ def get_text_attr(d, attr):
...
+        self._sort_formats(formats)
+
+        basic_url = "/".join(url.rstrip('/').split('/')[:3])
+        for f in formats:
+            f.setdefault('http_headers', {})['referer'] = basic_url + "/"
+            f.setdefault('http_headers', {})['origin'] = basic_url
+            f.setdefault('http_headers', {})['accept'] = "*/*"
+            f.setdefault('http_headers', {})['sec-fetch-dest'] = "empty"
+            f.setdefault('http_headers', {})['sec-fetch-mode'] = "cors"
+            f.setdefault('http_headers', {})['sec-fetch-site'] = "cross-site"
```
Comment (MinePlayersPE) [inline yt_dlp/extractor/cspan.py:278]:
```suggestion
            f.setdefault('http_headers', {}).update({
                'Referer': basic_url + '/',
                'Origin': basic_url,
                'Accept': '*/*',
                'Sec-Fetch-Dest': 'empty',
                'Sec-Fetch-Mode': 'cors',
                'Sec-Fetch-Site': 'cross-site'
            })
```

File: yt_dlp/extractor/cspan.py [hunk 3]
```diff
@@ -242,3 +243,49 @@ def get_text_attr(d, attr):
...
+        }
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._generic_id(url)
+        webpage = self._download_webpage(url, display_id)
+        file = open("sample.txt", "w")
+        file.write(webpage)
+        file.close()
+        video_url = re.findall(r'sources: \[\{file: \'(.*?)\'\}\],', webpage)[0]
```
Comment (MinePlayersPE) [inline yt_dlp/extractor/cspan.py:267]:
```suggestion
        video_url = self._search_regex(r'sources: \[\{file: \'(.*?)\'\}\],', webpage, 'video URL')
```
You can also just extract the jwplayer config with
```py
self._parse_json(
    self._search_regex(r'jwsetup\s*=\s*({.+});', webpage, 'player config'),
    video_id, transform_source=js_to_json),
```
 and use `_parse_jwplayer_data` to parse that (which should also extract captions), but ¯\\\_(ツ)_/¯

File: yt_dlp/extractor/cspan.py [hunk 4]
```diff
@@ -242,3 +243,49 @@ def get_text_attr(d, attr):
                 'title': title,
                 'id': 'c' + video_id if video_type == 'clip' else video_id,
             }
+
+
+class CSpanCongressIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?c-span\.org/congress/'
+    IE_DESC = 'C-SPAN'
```
Comment (MinePlayersPE) [inline yt_dlp/extractor/cspan.py:250]: imo this isn't descriptive, especially since it's copied from CSpanIE
```suggestion
```

File: yt_dlp/extractor/cspan.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/cspan.py:278]: Most of these headers are probably unnecessary too

File: yt_dlp/extractor/cspan.py [hunk 5]
```diff
@@ -242,3 +243,49 @@ def get_text_attr(d, attr):
...
+    _TESTS = [{
+        'url': 'https://www.c-span.org/congress/?chamber=house&date=2017-12-13&t=1513208380',
+        'info_dict': {
+            'id': 'Congressional Chronicle - Members of Congress, Hearings and More | C-SPAN.org',
```
Comment (Ashish0804) [inline yt_dlp/extractor/cspan.py:254]: i don't like this id tbh.
I personally would have parsed the `chamber` `date` and `t` from the url and maybe hashed them if not just joined them  using `_` and used that as the id.

File: yt_dlp/extractor/cspan.py [hunk 6]
```diff
@@ -242,3 +243,49 @@ def get_text_attr(d, attr):
...
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._generic_id(url)
+        webpage = self._download_webpage(url, display_id)
+        file = open("sample.txt", "w")
+        file.write(webpage)
+        file.close()
+        video_url = re.findall(r'sources: \[\{file: \'(.*?)\'\}\],', webpage)[0]
+        formats = self._extract_m3u8_formats(video_url, display_id, ext='mp4')
```
Comment (Ashish0804) [inline yt_dlp/extractor/cspan.py:268]:
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(video_url, display_id, ext='mp4')
```

File: yt_dlp/extractor/cspan.py (hunk 2 above)
Comment (Ashish0804) [reply yt_dlp/extractor/cspan.py:278]: Also u can just hardcode the `basic_url` to be `'https://www.c-span.org'`
And i'm pretty sure having `/` in the end for origin aswell won't break things so u can have the basic_url as 
`'https://www.c-span.org'/`

Comment (Lesmiscore) [reply yt_dlp/extractor/cspan.py:278]: I don't know whether this website does, but sometimes `Origin` header is strictly checked

Comment (MinePlayersPE) [reply yt_dlp/extractor/cspan.py:278]:
```suggestion
            f.setdefault('http_headers', {})['Referer'] = 'https://www.c-span.org/'
```
Not directly tested, but seems like only `Referer` is actually needed

File: yt_dlp/extractor/cspan.py (hunk 5 above)
Comment (MinePlayersPE) [reply yt_dlp/extractor/cspan.py:254]: +1, the webpage title seems static across all videos so this would completely break archiving

**Comment by Grabien** [comment]
I added the necessary corrections. I hope it looks better now.

**Comment by Grabien** [comment]
Also added a small fix for wrong title code, like this: "<h3 class='video-page-title' itemprop='name'>U.S. Senate</h2>"
It's found on this page:
https://www.c-span.org/video/?517228-1/senate-debates-russian-sanctions-voting-rights-legislation

File: yt_dlp/extractor/cspan.py [hunk 7]
```diff
@@ -1,6 +1,10 @@
 from __future__ import unicode_literals
 
 import re
+from urllib.parse import (
+    parse_qs,
+    urlparse,
+)
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:7]:
```suggestion
import urllib
```

File: yt_dlp/extractor/cspan.py [hunk 8]
```diff
@@ -126,8 +130,11 @@ def add_referer(formats):
                         ext = 'vtt'
                     subtitle['ext'] = ext
             ld_info = self._search_json_ld(webpage, video_id, default={})
-            title = get_element_by_class('video-page-title', webpage) or \
-                self._og_search_title(webpage)
+            try:
+                title = get_element_by_class('video-page-title', webpage) or \
+                    self._og_search_title(webpage)
+            except Exception:
+                title = 'No title'
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:137]: Why?

File: yt_dlp/extractor/cspan.py [hunk 9]
```diff
@@ -180,6 +187,7 @@ def get_text_attr(d, attr):
         data = self._download_json(
             'http://www.c-span.org/assets/player/ajax-player.php?os=android&html5=%s&id=%s' % (video_type, video_id),
             video_id)['video']
+
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:190]:
```suggestion
```
unrelated diff

File: yt_dlp/extractor/cspan.py [hunk 10]
```diff
@@ -242,3 +250,42 @@ def get_text_attr(d, attr):
...
+class CSpanCongressIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?c-span\.org/congress/'
+    _TESTS = [{
+        'url': 'https://www.c-span.org/congress/?chamber=house&date=2017-12-13&t=1513208380',
+        'info_dict': {
+            'id': 'house_2017-12-13',
+            'title': 'Congressional Chronicle - Members of Congress, Hearings and More | C-SPAN.org',
+            'description': 'md5:54c264b7a8f219937987610243305a84',
+            'thumbnail': 'https://ximage.c-spanvideo.org/eyJidWNrZXQiOiJwaWN0dXJlcy5jLXNwYW52aWRlby5vcmciLCJrZXkiOiJGaWxlc1wvODdkXC8yMDE3MTIxMzEwMDAzMjAwMV9oZC5qcGciLCJlZGl0cyI6eyJyZXNpemUiOnsiZml0IjoiY292ZXIiLCJoZWlnaHQiOjU3Niwid2lkdGgiOjEwMjR9fX0=',
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:263]:
```suggestion
            'thumbnail': r're:https://ximage.c-spanvideo.org/.+',
```

File: yt_dlp/extractor/cspan.py [hunk 11]
```diff
@@ -242,3 +250,42 @@ def get_text_attr(d, attr):
...
             }
+
+
+class CSpanCongressIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?c-span\.org/congress/'
+    _TESTS = [{
+        'url': 'https://www.c-span.org/congress/?chamber=house&date=2017-12-13&t=1513208380',
+        'info_dict': {
+            'id': 'house_2017-12-13',
+            'title': 'Congressional Chronicle - Members of Congress, Hearings and More | C-SPAN.org',
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:261]: `| C-SPAN.org` should be removed

File: yt_dlp/extractor/cspan.py [hunk 12]
```diff
@@ -242,3 +250,42 @@ def get_text_attr(d, attr):
...
+            'description': 'md5:54c264b7a8f219937987610243305a84',
+            'thumbnail': 'https://ximage.c-spanvideo.org/eyJidWNrZXQiOiJwaWN0dXJlcy5jLXNwYW52aWRlby5vcmciLCJrZXkiOiJGaWxlc1wvODdkXC8yMDE3MTIxMzEwMDAzMjAwMV9oZC5qcGciLCJlZGl0cyI6eyJyZXNpemUiOnsiZml0IjoiY292ZXIiLCJoZWlnaHQiOjU3Niwid2lkdGgiOjEwMjR9fX0=',
+            'ext': 'mp4'
+        }
+    }]
+
+    def _real_extract(self, url):
+        parsed_url = urlparse(url)
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:269]:
```suggestion
        query = parse_qs(url)
```
Use the `parse_qs` from utils

File: yt_dlp/extractor/cspan.py [hunk 13]
```diff
@@ -242,3 +250,42 @@ def get_text_attr(d, attr):
...
+        if parse_qs(parsed_url.query)['chamber'] is not None:
+            display_id = parse_qs(parsed_url.query)['chamber'][0]
+            if parse_qs(parsed_url.query)['date'] is not None:
+                display_id = display_id + '_' + parse_qs(parsed_url.query)['date'][0]
+        else:
+            display_id = self._generic_id(url)
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:275]: add test for this case

File: yt_dlp/extractor/cspan.py [hunk 14]
```diff
@@ -242,3 +250,42 @@ def get_text_attr(d, attr):
...
+
+        for f in result['formats']:
+            f.setdefault('http_headers', {})['Referer'] = 'https://www.c-span.org/'
+
+        result['title'] = self._og_search_title(webpage, default=None) or self._html_search_regex(
+            r'(?s)<title>(.*?)</title>', webpage, 'video title')
+        result['description'] = self._og_search_description(webpage, default=None) or self._html_search_meta(
+            'description', webpage, 'meta description', default=None)
+
+        return result
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:291]:
```suggestion
        title = (self._og_search_title(webpage, default=None)
                 or self._html_search_regex(r'(?s)<title>(.*?)</title>', webpage, 'video title'))
        description = (self._og_search_description(webpage, default=None)
                       or self._html_search_meta('description', webpage, 'description', default=None))

        return {
            **self._parse_jwplayer_data(jwplayer_data, display_id, False),
            'title': title,
            'description': description,
            'http_headers': {'Referer': 'https://www.c-span.org/'},
        }
```

File: yt_dlp/extractor/cspan.py (hunk 8 above)
Comment (Grabien) [reply yt_dlp/extractor/cspan.py:134]: Some C-Span website pages have an incorrect closing H tag, like this:
`<h3 class='video-page-title' itemprop='name'>U.S. Senate</h2>`

Because of this, the file can't be downloaded with this error message:
`ERROR: matching opening tag for closing h2 tag not found`

It's a workaround for such issues.

**Comment by Grabien** [comment]
All necessary corrections are added.

File: yt_dlp/extractor/cspan.py (hunk 8 above)
Comment (pukkandan) [reply yt_dlp/extractor/cspan.py:134]: example?

File: yt_dlp/extractor/cspan.py [hunk 15]
```diff
@@ -242,3 +246,52 @@ def get_text_attr(d, attr):
...
+    }, {
+        'url': 'https://www.c-span.org/congress/',
+        'info_dict': {
+            'id': 'congress',
+            'title': 'Congressional Chronicle - Members of Congress, Hearings and More',
+            'description': 'md5:54c264b7a8f219937987610243305a84',
+            'thumbnail': r're:https://ximage.c-spanvideo.org/.+',
+            'ext': 'mp4'
+        }
+    }]
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:271]: This is a problem. The video id MUST be both unique and consistent (per extractor). ie, (1) Diff videos must never have the same id and, (2) A video must always have the same id even if download using different URLs. This does not seem to satisfy either case

Comment (Grabien) [reply yt_dlp/extractor/cspan.py:271]: Regarding the first question:

```
yt-dlp "https://www.c-span.org/video/?517228-1/senate-debates-russian-sanctions-voting-rights-legislation" -v

[debug] Command-line config: ['https://www.c-span.org/video/?517228-1/senate-debates-russian-sanctions-voting-rights-legislation', '-v']
[debug] Encodings: locale cp1251, fs utf-8, out utf-8, err utf-8, pref cp1251
[debug] yt-dlp version 2021.12.27 [6223f67a8]
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Python version 3.8.7 (CPython 64bit) - Windows-10-10.0.19041-SP0
[debug] exe versions: ffmpeg 4.4-full_build-www.gyan.dev (setts), ffprobe 4.4-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
[debug] [CSpan] Extracting URL: https://www.c-span.org/video/?517228-1/senate-debates-russian-sanctions-voting-rights-legislation
[CSpan] 517228: Downloading webpage
[CSpan] 517228: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
ERROR: matching opening tag for closing h2 tag not found
Traceback (most recent call last):
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\YoutubeDL.py", line 1365, in wrapper
    return func(self, *args, **kwargs)
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\YoutubeDL.py", line 1435, in __extract_info
    ie_result = ie.extract(url)
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\extractor\common.py", line 609, in extract
    ie_result = self._real_extract(url)
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\extractor\cspan.py", line 129, in _real_extract
    title = get_element_by_class('video-page-title', webpage) or \
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\utils.py", line 1964, in get_element_by_class
    retval = get_elements_by_class(class_name, html)
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\utils.py", line 1986, in get_elements_by_class
    return get_elements_by_attribute(
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\utils.py", line 2000, in get_elements_by_attribute
    return [content for content, _ in get_elements_text_and_html_by_attribute(*args, **kwargs)]
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\utils.py", line 2000, in <listcomp>
    return [content for content, _ in get_elements_text_and_html_by_attribute(*args, **kwargs)]
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\utils.py", line 2025, in get_elements_text_and_html_by_attribute
    content, whole = get_element_text_and_html_by_tag(m.group('tag'), html[m.start():])
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\utils.py", line 2102, in get_element_text_and_html_by_tag
    parser.feed(html[offset:offset + next_closing_tag_end])
  File "c:\program files\python38\lib\html\parser.py", line 111, in feed
    self.goahead(0)
  File "c:\program files\python38\lib\html\parser.py", line 173, in goahead
    k = self.parse_endtag(i)
  File "c:\program files\python38\lib\html\parser.py", line 421, in parse_endtag
    self.handle_endtag(elem)
  File "c:\work\projects\grabien\yt-dlp\yt_dlp\utils.py", line 2070, in handle_endtag
    raise compat_HTMLParseError(f'matching opening tag for closing {tag} tag not found')
yt_dlp.compat.compat_HTMLParseError: matching opening tag for closing h2 tag not found
```

File: yt_dlp/extractor/cspan.py (hunk 8 above)
Comment (pukkandan) [reply yt_dlp/extractor/cspan.py:134]:
```suggestion
            title = get_element_by_class('video-page-title', webpage) or self._og_search_title(webpage)
```

Merge the PR with master and this will work

File: yt_dlp/extractor/cspan.py [hunk 16]
```diff
@@ -242,3 +246,49 @@ def get_text_attr(d, attr):
...
+            video_id = query['chamber'][0]
+        else:
+            video_id = 'senate'
+        if 'date' in query:
+            video_date = query['date'][0]
+            video_id = video_id + '_' + video_date
+        else:
+            video_date = ''
+        webpage = self._download_webpage(url, video_id)
+        if video_date == '':
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:276]:
```suggestion
        video_date = query.get('date', [None])[0]
        video_id = join_nonempty(query.get('chamber', ['senate'])[0], video_date, delim='_')
        webpage = self._download_webpage(url, video_id)
        if not video_date:
```

File: yt_dlp/extractor/cspan.py [hunk 17]
```diff
@@ -242,3 +246,49 @@ def get_text_attr(d, attr):
...
+            jwp_date = re.search(r'jwsetup.clipprogdate = \'(?P<date>\d{4}-\d{2}-\d{2})\';', webpage)
+            if jwp_date:
+                video_id = video_id + '_' + jwp_date.group('date')
```
Comment (pukkandan) [inline yt_dlp/extractor/cspan.py:279]: avoid `+` ing strings. Use fstrings/%s
```suggestion
                video_id = f'{video_id}_{jwp_date.group("date")}'
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Review by pukkandan — APPROVED** [review]
pls fix the import and check I havent broken anything

File: yt_dlp/extractor/cspan.py (hunk 8 above)
Comment (Grabien) [reply yt_dlp/extractor/cspan.py]: > Merge the PR with master and this will work

I merged, but it still did not work. The same error:
`yt_dlp.compat.compat_HTMLParseError: matching opening tag for closing h2 tag not found`

**Comment by Grabien** [comment]
Imports are fixed. Everything works fine, except for the bug with the wrong H2 tag.

File: yt_dlp/extractor/cspan.py (hunk 8 above)
Comment (pukkandan) [reply yt_dlp/extractor/cspan.py]: right, my bad. But catch only the relevant exception

**Comment by Grabien** [comment]
I added the necessary corrections.
